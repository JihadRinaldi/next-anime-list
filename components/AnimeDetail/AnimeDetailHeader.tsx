import { AnimeResponse } from '@/types/animeServiceType';

import { ImageWithFallback } from '../ImageWithFallback';
import { Text } from '../Text';

type AnimeDetailProps = {
  animeData: AnimeResponse;
};

const AnimeDetailHeader = ({ animeData }: AnimeDetailProps) => {
  const getGenres = animeData.genres.map((genre) => genre.name).join(', ');
  const animeStatisticsClassName = 'text-3xl font-bold';

  const statisticsTitle = (title: string) => (
    <Text
      label={title}
      className="w-full text-center bg-blue-600 px-2 rounded-md mb-1"
    />
  );

  return (
    <div className="p-4 flex gap-2 w-full bg-eerie-black rounded-md">
      <div className="relative aspect-[184/289] w-48 h-full">
        <ImageWithFallback
          fill
          src={animeData.images.webp.image_url}
          alt={animeData.title}
          className="object-cover"
        />
      </div>
      <div className="p-4 flex flex-col gap-4">
        <h1 className="text-3xl font-bold">{animeData.title}</h1>
        <div className="flex flex-col gap-4 justify-between">
          <div className="flex gap-6 items-start">
            {animeData.score && animeData.scored_by ? (
              <div className="flex flex-col items-center justify-center gap-1">
                {statisticsTitle('SCORE')}
                <Text
                  label={`${animeData.score}`}
                  className={animeStatisticsClassName}
                />
                <Text
                  className="text-center"
                  label={`${animeData.scored_by.toLocaleString()} users`}
                />
              </div>
            ) : null}
            {animeData.rank ? (
              <div className="flex flex-col items-center justify-center gap-1">
                {statisticsTitle('RANKED')}
                <Text
                  label={`#${animeData.rank}`}
                  className={animeStatisticsClassName}
                />
              </div>
            ) : null}
            <div className="flex flex-col items-center justify-center gap-1">
              {statisticsTitle('POPULARITY')}
              <Text
                label={`#${animeData.popularity}`}
                className={animeStatisticsClassName}
              />
            </div>
            <div className="flex flex-col items-center justify-center gap-1">
              {statisticsTitle('MEMBERS')}
              <Text
                label={animeData.members.toLocaleString()}
                className={animeStatisticsClassName}
              />
            </div>
            <div className="flex flex-col items-center justify-center gap-1">
              {statisticsTitle('FAVORITES')}
              <Text
                label={animeData.favorites.toLocaleString()}
                className={animeStatisticsClassName}
              />
            </div>
          </div>
          <div className="flex gap-4">
            <div className="flex flex-col gap-2">
              <div className="flex gap-2">
                <Text className="font-semibold" label="Status: " />
                <Text label={animeData.status} />
              </div>
              <div className="flex gap-2">
                <Text className="font-semibold" label="Episodes: " />
                <Text label={`${animeData.episodes}`} />
              </div>
              <div className="flex gap-2">
                <Text className="font-semibold" label="Type: " />
                <Text label={animeData.type} />
              </div>
            </div>
            <div className="flex flex-col gap-2">
              {animeData.genres.length ? (
                <div className="flex gap-2">
                  <Text className="font-semibold" label="Genres: " />
                  <Text label={getGenres} />
                </div>
              ) : null}
              <div className="flex gap-2">
                <Text className="font-semibold" label="Duration: " />
                <Text label={animeData.duration} />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default AnimeDetailHeader;
