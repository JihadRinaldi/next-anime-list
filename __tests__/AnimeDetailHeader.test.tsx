import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';

import AnimeDetailHeader from '@/components/AnimeDetail/AnimeDetailHeader';

import {
  MOCK_SINGLE_ITEM,
  MOCK_SINGLE_ITEM_WITH_NULL_DATA,
} from '../__mocks__/anime';

describe('AnimeDetailHeader with null data', () => {
  it('renders the anime title', () => {
    render(<AnimeDetailHeader animeData={MOCK_SINGLE_ITEM_WITH_NULL_DATA} />);
    expect(screen.getByText('Dokkiri Doctor Special')).toBeInTheDocument();
  });

  it('does not render the score and scored_by when they are null', () => {
    render(<AnimeDetailHeader animeData={MOCK_SINGLE_ITEM_WITH_NULL_DATA} />);
    expect(screen.queryByText('SCORE')).not.toBeInTheDocument();
    expect(screen.queryByText('0 users')).not.toBeInTheDocument();
  });

  it('renders the rank', () => {
    render(<AnimeDetailHeader animeData={MOCK_SINGLE_ITEM_WITH_NULL_DATA} />);
    expect(screen.getByText('#19340')).toBeInTheDocument();
  });

  it('renders the popularity, members, and favorites', () => {
    render(<AnimeDetailHeader animeData={MOCK_SINGLE_ITEM_WITH_NULL_DATA} />);
    expect(screen.getByText('#16510')).toBeInTheDocument();
    expect(screen.getByText('467')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
  });

  it('renders the status, episodes, type, genres, and duration', () => {
    render(<AnimeDetailHeader animeData={MOCK_SINGLE_ITEM_WITH_NULL_DATA} />);
    expect(screen.getByText('Status:')).toBeInTheDocument();
    expect(screen.getByText('Finished Airing')).toBeInTheDocument();
    expect(screen.getByText('Episodes:')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('Type:')).toBeInTheDocument();
    expect(screen.getByText('Special')).toBeInTheDocument();
    expect(screen.getByText('Genres:')).toBeInTheDocument();
    expect(screen.getByText('Comedy, Ecchi')).toBeInTheDocument();
    expect(screen.getByText('Duration:')).toBeInTheDocument();
    expect(screen.getByText('25 min')).toBeInTheDocument();
  });

  it('renders the image with correct src and alt text', () => {
    render(<AnimeDetailHeader animeData={MOCK_SINGLE_ITEM_WITH_NULL_DATA} />);
    const image = screen.getByAltText('Dokkiri Doctor Special');
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src');
  });
});

describe('AnimeDetailHeader Snapshot Tests', () => {
  it('renders correctly with full data', () => {
    const { container } = render(
      <AnimeDetailHeader animeData={MOCK_SINGLE_ITEM} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with partial data', () => {
    const { container } = render(
      <AnimeDetailHeader animeData={MOCK_SINGLE_ITEM_WITH_NULL_DATA} />
    );
    expect(container).toMatchSnapshot();
  });
});
