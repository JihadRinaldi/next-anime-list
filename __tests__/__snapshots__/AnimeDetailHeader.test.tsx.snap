// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimeDetailHeader Snapshot Tests renders correctly with full data 1`] = `
<div>
  <div
    class="p-4 flex gap-2 w-full bg-eerie-black rounded-md"
  >
    <div
      class="relative aspect-[184/289] w-48 h-full"
    >
      <img
        alt="Dokkiri Doctor Special"
        class="object-cover"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="http://localhost/_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="p-4 flex flex-col gap-4"
    >
      <h1
        class="text-3xl font-bold"
      >
        Dokkiri Doctor Special
      </h1>
      <div
        class="flex flex-col gap-4 justify-between"
      >
        <div
          class="flex gap-6 items-start"
        >
          <div
            class="flex flex-col items-center justify-center gap-1"
          >
            <div
              class="text-white w-full text-center bg-blue-600 px-2 rounded-md mb-1"
            >
              SCORE
            </div>
            <div
              class="text-white text-3xl font-bold"
            >
              8.5
            </div>
            <div
              class="text-white text-center"
            >
              1,500 users
            </div>
          </div>
          <div
            class="flex flex-col items-center justify-center gap-1"
          >
            <div
              class="text-white w-full text-center bg-blue-600 px-2 rounded-md mb-1"
            >
              RANKED
            </div>
            <div
              class="text-white text-3xl font-bold"
            >
              #19340
            </div>
          </div>
          <div
            class="flex flex-col items-center justify-center gap-1"
          >
            <div
              class="text-white w-full text-center bg-blue-600 px-2 rounded-md mb-1"
            >
              POPULARITY
            </div>
            <div
              class="text-white text-3xl font-bold"
            >
              #16510
            </div>
          </div>
          <div
            class="flex flex-col items-center justify-center gap-1"
          >
            <div
              class="text-white w-full text-center bg-blue-600 px-2 rounded-md mb-1"
            >
              MEMBERS
            </div>
            <div
              class="text-white text-3xl font-bold"
            >
              467
            </div>
          </div>
          <div
            class="flex flex-col items-center justify-center gap-1"
          >
            <div
              class="text-white w-full text-center bg-blue-600 px-2 rounded-md mb-1"
            >
              FAVORITES
            </div>
            <div
              class="text-white text-3xl font-bold"
            >
              2
            </div>
          </div>
        </div>
        <div
          class="flex gap-4"
        >
          <div
            class="flex flex-col gap-2"
          >
            <div
              class="flex gap-2"
            >
              <div
                class="text-white font-semibold"
              >
                Status: 
              </div>
              <div
                class="text-white"
              >
                Finished Airing
              </div>
            </div>
            <div
              class="flex gap-2"
            >
              <div
                class="text-white font-semibold"
              >
                Episodes: 
              </div>
              <div
                class="text-white"
              >
                1
              </div>
            </div>
            <div
              class="flex gap-2"
            >
              <div
                class="text-white font-semibold"
              >
                Type: 
              </div>
              <div
                class="text-white"
              >
                Special
              </div>
            </div>
          </div>
          <div
            class="flex flex-col gap-2"
          >
            <div
              class="flex gap-2"
            >
              <div
                class="text-white font-semibold"
              >
                Genres: 
              </div>
              <div
                class="text-white"
              >
                Comedy, Ecchi
              </div>
            </div>
            <div
              class="flex gap-2"
            >
              <div
                class="text-white font-semibold"
              >
                Duration: 
              </div>
              <div
                class="text-white"
              >
                25 min
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AnimeDetailHeader Snapshot Tests renders correctly with partial data 1`] = `
<div>
  <div
    class="p-4 flex gap-2 w-full bg-eerie-black rounded-md"
  >
    <div
      class="relative aspect-[184/289] w-48 h-full"
    >
      <img
        alt="Dokkiri Doctor Special"
        class="object-cover"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="http://localhost/_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fcdn.myanimelist.net%2Fimages%2Fanime%2F12%2F44077.webp&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="p-4 flex flex-col gap-4"
    >
      <h1
        class="text-3xl font-bold"
      >
        Dokkiri Doctor Special
      </h1>
      <div
        class="flex flex-col gap-4 justify-between"
      >
        <div
          class="flex gap-6 items-start"
        >
          <div
            class="flex flex-col items-center justify-center gap-1"
          >
            <div
              class="text-white w-full text-center bg-blue-600 px-2 rounded-md mb-1"
            >
              RANKED
            </div>
            <div
              class="text-white text-3xl font-bold"
            >
              #19340
            </div>
          </div>
          <div
            class="flex flex-col items-center justify-center gap-1"
          >
            <div
              class="text-white w-full text-center bg-blue-600 px-2 rounded-md mb-1"
            >
              POPULARITY
            </div>
            <div
              class="text-white text-3xl font-bold"
            >
              #16510
            </div>
          </div>
          <div
            class="flex flex-col items-center justify-center gap-1"
          >
            <div
              class="text-white w-full text-center bg-blue-600 px-2 rounded-md mb-1"
            >
              MEMBERS
            </div>
            <div
              class="text-white text-3xl font-bold"
            >
              467
            </div>
          </div>
          <div
            class="flex flex-col items-center justify-center gap-1"
          >
            <div
              class="text-white w-full text-center bg-blue-600 px-2 rounded-md mb-1"
            >
              FAVORITES
            </div>
            <div
              class="text-white text-3xl font-bold"
            >
              2
            </div>
          </div>
        </div>
        <div
          class="flex gap-4"
        >
          <div
            class="flex flex-col gap-2"
          >
            <div
              class="flex gap-2"
            >
              <div
                class="text-white font-semibold"
              >
                Status: 
              </div>
              <div
                class="text-white"
              >
                Finished Airing
              </div>
            </div>
            <div
              class="flex gap-2"
            >
              <div
                class="text-white font-semibold"
              >
                Episodes: 
              </div>
              <div
                class="text-white"
              >
                1
              </div>
            </div>
            <div
              class="flex gap-2"
            >
              <div
                class="text-white font-semibold"
              >
                Type: 
              </div>
              <div
                class="text-white"
              >
                Special
              </div>
            </div>
          </div>
          <div
            class="flex flex-col gap-2"
          >
            <div
              class="flex gap-2"
            >
              <div
                class="text-white font-semibold"
              >
                Genres: 
              </div>
              <div
                class="text-white"
              >
                Comedy, Ecchi
              </div>
            </div>
            <div
              class="flex gap-2"
            >
              <div
                class="text-white font-semibold"
              >
                Duration: 
              </div>
              <div
                class="text-white"
              >
                25 min
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
